Quirk getElementsByName() on older iPadOS apps

annevk opened the pull request:
#### 290968e36d41b424e268d39e9a2774c64cc6bd87
<pre>
Quirk getElementsByName() on older iPadOS apps
<a href="https://bugs.webkit.org/show_bug.cgi?id=270229">https://bugs.webkit.org/show_bug.cgi?id=270229</a>
<a href="https://rdar.apple.com/122548304">rdar://122548304</a>

Reviewed by Ryosuke Niwa.

272530@main aligned us with the HTML Standard for getElementsByName()
by making it only work for HTML elements.

There is at least one iPadOS app that is impacted by that change. This
change should give them a bit of time to catch up by making
getElementsByName() look at all elements again for older iPadOS apps.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/dom/NameNodeList.cpp:
(WebCore::NameNodeList::NameNodeList):
(WebCore::NameNodeList::elementMatches const):
* Source/WebCore/dom/NameNodeList.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGetElementsByNameQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/275779@main">https://commits.webkit.org/275779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2c92cbf30a4c09f03f9f3218f1862379c906f87

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21843 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45436 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38947 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19223 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18860 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36857 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16398 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16476 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37921 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/878 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/36278 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38990 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38255 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46948 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42449 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17649 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14544 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19274 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40806 "Found 1 new API test failure: /WebKitGTK/TestWebKitFaviconDatabase:/webkit/WebKitFaviconDatabase/get-favicon (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9549 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19438 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49459 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18904 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9996 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/NameNodeList.cpp:
```diff
@@ -52,7 +53,7 @@ NameNodeList::~NameNodeList()
 bool NameNodeList::elementMatches(Element& element) const
 {
-    return element.isHTMLElement() && element.getNameAttribute() == m_name;
+    return (is<HTMLElement>(element) || element.document().quirks().needsGetElementsByNameQuirk()) && element.getNameAttribute() == m_name;
```
I don't think we want to be checking the quirk value for every element like this.
We should probably cache the value in a member variable so that we don't do a function call & two indirect loads for every element.

annevk replied:
Done. Not entirely sure if the const_cast is avoidable. But I hope that we can remove this code within a year.

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1838,4 +1837,15 @@ bool Quirks::shouldIgnorePlaysInlineRequirementQuirk() const
 }
 
+// This quirk has intentionally limited exposure to increase the odds of being able to remove it
+// again within a reasonable timeframe as the impacted apps are being updated.
```
reference to the radar would be useful for readers/testers in the future.

rniwa commented on Source/WebCore/dom/NameNodeList.cpp:
```diff
@@ -50,9 +51,16 @@ NameNodeList::~NameNodeList()
 
+bool NameNodeList::needsGetElementsByNameQuirk()
+{
+    if (m_needsGetElementsByNameQuirk == NeedsGetElementsByNameQuirk::Unset)
```
We should do this initialization in the constructor instead.

rniwa reviewed (changes requested):
I don't think it's acceptable to add a function call in the middle of elementMatches like this.

annevk committed:
Quirk getElementsByName() on older iPadOS apps
https://bugs.webkit.org/show_bug.cgi?id=270229
rdar://122548304

Reviewed by Ryosuke Niwa.

272530@main aligned us with the HTML Standard for getElementsByName()
by making it only work for HTML elements.

There is at least one iPadOS app that is impacted by that change. This
change should give them a bit of time to catch up by making
getElementsByName() look at all elements again for older iPadOS apps.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/dom/NameNodeList.cpp:
(WebCore::NameNodeList::NameNodeList):
(WebCore::NameNodeList::elementMatches const):
* Source/WebCore/dom/NameNodeList.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGetElementsByNameQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/275779@main
